src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -1055,16 +1055,71 @@ public void Connect(IPAddress[] addresses, int port)
                 throw new NotSupportedException(SR.net_invalidversion);
             }
 
+// Disable CS0162: Unreachable code detected
+//
+// SuportsMultipleConnectAttempts is a constant; when false, the following lines will trigger CS0162.
+#pragma warning disable 162
             Exception lastex = null;
-            foreach (IPAddress address in addresses)
+            if (SocketPal.SupportsMultipleConnectAttempts)
             {
-                if (CanTryAddressFamily(address.AddressFamily))
+                foreach (IPAddress address in addresses)
+                {
+                    if (CanTryAddressFamily(address.AddressFamily))
+                    {
+                        try
+                        {
+                            Connect(new IPEndPoint(address, port));
+                            lastex = null;
+                            break;
+                        }
+                        catch (Exception ex)
+                        {
+                            if (ExceptionCheck.IsFatal(ex)) throw;
+                            lastex = ex;
+                        }
+                    }
+                }
+            }
+            else
+            {
+                EndPoint endpoint = null;
+                foreach (IPAddress address in addresses)
+                {
+                    if (CanTryAddressFamily(address.AddressFamily))
+                    {
+                        Socket attemptSocket = null;
+                        try
+                        {
+                            attemptSocket = new Socket(_addressFamily, _socketType, _protocolType);
+                            if (IsDualMode)
+                            {
+                                attemptSocket.DualMode = true;
+                            }
+
+                            var attemptEndpoint = new IPEndPoint(address, port);
+                            attemptSocket.Connect(attemptEndpoint);
+                            attemptSocket.Dispose();
```

Comment:
You might consider adding a finally that disposes of attemptSocket rather than doing it both here and in the catch.